### PR TITLE
Create paths for external repositories in a way that is forward and backward compatible

### PR DESCRIPTION
**Pull request by kchodorow**
Bazel has, up until this point, put external repositories "under" the local repository (e.g., execroot/local_repo/bazel-out/local-fastbuild/genfiles/external/remote_repo/path/to/target).  However, an upcoming CL is changing that so that all repositories are "siblings:" remote_repo would be at execroot/remote_repo/bazel-out/local-fastbuild/genfiles/path/to/target.  This CL makes the proto rule use the right path for both the old-style file structure and the new-style.

Most of this complexity can be removed once new-style is standard.

### CODE REVIEW

**Comment by kchodorow** [comment]
Friendly ping?

**Comment by xfxyjwf** [comment]
@kchodorow Can you ask someone from bazel team to review the change? I'm not very familiar with bazel details but would be happy to help merge the change as long as someone from bazel team says LGTM.

**Comment by kchodorow** [comment]
Sure, thanks!

File: protobuf.bzl [hunk 1]
```diff
@@ -1,19 +1,56 @@
...
+        parts += [genfiles_fragment]
+      parts += [ctx.label.workspace_root]
+      if suffix:
+        parts += [suffix]
+    else:
+      parts += [ctx.label.workspace_root]
+      if genfiles_fragment:
+        parts += [genfiles_fragment]
+      if suffix:
+        parts += [suffix]
```
Comment (damienmg) [inline protobuf.bzl]: Factor out the if suffix part

**Comment by dslomov** [comment]
Looks ok Windows-wise

File: protobuf.bzl (hunk 1 above)
Comment (kchodorow) [inline protobuf.bzl]: Done.

**Comment by kchodorow** [comment]
Unfortunately, the tests don't pass for me locally (with or without my change, old or new version of Bazel) so it's hard for me to verify that this won't break ci.bazel.io.  But... it shouldn't?

I get:

```
INFO: Elapsed time: 2.624s, Critical Path: 2.06s
INFO: Build completed, 8 tests FAILED, 21 total actions
//:protobuf_test                                                (cached) PASSED in 10.0s
//:py_descriptor_database_test                                  (cached) PASSED in 0.1s
//:py_message_factory_test                                      (cached) PASSED in 0.1s
//:py_proto_builder_test                                        (cached) PASSED in 0.1s
//:py_service_reflection_test                                   (cached) PASSED in 0.1s
//:py_symbol_database_test                                      (cached) PASSED in 0.1s
//:py_text_encoding_test                                        (cached) PASSED in 0.1s
//:py_wire_format_test                                          (cached) PASSED in 0.1s
//:py_descriptor_pool_test                                               FAILED in 0.3s
  /usr/local/google/home/kchodorow/.cache/bazel/_bazel_kchodorow/6336bcba2f023ab45d699b44e835c9a6/execroot/protobuf/bazel-out/local-fastbuild/testlogs/py_descriptor_pool_test/test.log
//:py_descriptor_test                                                    FAILED in 0.3s
  /usr/local/google/home/kchodorow/.cache/bazel/_bazel_kchodorow/6336bcba2f023ab45d699b44e835c9a6/execroot/protobuf/bazel-out/local-fastbuild/testlogs/py_descriptor_test/test.log
//:py_generator_test                                                     FAILED in 0.7s
  /usr/local/google/home/kchodorow/.cache/bazel/_bazel_kchodorow/6336bcba2f023ab45d699b44e835c9a6/execroot/protobuf/bazel-out/local-fastbuild/testlogs/py_generator_test/test.log
//:py_json_format_test                                                   FAILED in 0.7s
  /usr/local/google/home/kchodorow/.cache/bazel/_bazel_kchodorow/6336bcba2f023ab45d699b44e835c9a6/execroot/protobuf/bazel-out/local-fastbuild/testlogs/py_json_format_test/test.log
//:py_message_test                                                       FAILED in 0.4s
  /usr/local/google/home/kchodorow/.cache/bazel/_bazel_kchodorow/6336bcba2f023ab45d699b44e835c9a6/execroot/protobuf/bazel-out/local-fastbuild/testlogs/py_message_test/test.log
//:py_reflection_test                                                    FAILED in 0.4s
  /usr/local/google/home/kchodorow/.cache/bazel/_bazel_kchodorow/6336bcba2f023ab45d699b44e835c9a6/execroot/protobuf/bazel-out/local-fastbuild/testlogs/py_reflection_test/test.log
//:py_text_format_test                                                   FAILED in 0.3s
  /usr/local/google/home/kchodorow/.cache/bazel/_bazel_kchodorow/6336bcba2f023ab45d699b44e835c9a6/execroot/protobuf/bazel-out/local-fastbuild/testlogs/py_text_format_test/test.log
//:py_unknown_fields_test                                                FAILED in 0.3s
  /usr/local/google/home/kchodorow/.cache/bazel/_bazel_kchodorow/6336bcba2f023ab45d699b44e835c9a6/execroot/protobuf/bazel-out/local-fastbuild/testlogs/py_unknown_fields_test/test.log

Executed 8 out of 16 tests: 8 tests pass and 8 fail locally.
There were tests whose specified size is too big. Use the --test_verbose_timeout_warnings command line option to see which ones these are.
```

**Comment by kchodorow** [comment]
Hey, remember this old thing?  Updated to work.  Passes all tests with new & old bazel.

PTAL!

**Comment by kchodorow** [comment]
Yeah, it's been a while.

@xfxyjwf, could you merge this in?

**Pull request merged**